src/D2L.CodeStyle.TestAnalyzers/ServiceLocator/CustomTestServiceLocatorAnalyzer.cs

```diff
@@ -0,0 +1,189 @@
+			if( root == null ) {
+				return;
+			}
+
+			SymbolInfo symbolInfo = context.SemanticModel.GetSymbolInfo( root );
+
+			IMethodSymbol method = symbolInfo.Symbol as IMethodSymbol;
+
+			if( method == null ) {
+				if( symbolInfo.CandidateSymbols == null ) {
+					return;
+				}
+
+				if( symbolInfo.CandidateSymbols.Length != 1 ) {
+					return;
+				}
+
+				// This happens on method groups, such as
+				// Func<IServiceLocator> fooFunc = TestServiceLocatorFactory.Create( ... );
+				method = symbolInfo.CandidateSymbols.First() as IMethodSymbol;
+
+				if( method == null ) {
+					return;
+				}
+			}
+
+			// If we're a Create method on a class that isn't
+			// TestServiceLocatorFactory, we're safe.
+			if( !IsTestServiceLocatorFactory(
+					actualType: method.ContainingType,
+					disallowedType: disallowedType
+				)
+			) {
+				return;
+			}
+
+			// If we're not a Create method, we're safe.
+			if( !IsCreateMethod( method ) ) {
+				return;
+			}
+
+			// Check whether any parent classes are on our whitelist. Checking
+			// all of the parents lets us handle partial classes more easily.
+			var parentClasses = context.Node.Ancestors().Where(
+				a => a.IsKind( SyntaxKind.ClassDeclaration )
+			);
+
+			var parentSymbols = parentClasses.Select(
+				c => context.SemanticModel.GetDeclaredSymbol( c )
+			);
+
+			bool isWhitelisted = parentSymbols.Any(
+				s => IsClassWhitelisted( whitelistedClasses, s )
+			);
+
+			if( isWhitelisted ) {
+				return;
+			}
+
+			context.ReportDiagnostic(
+				Diagnostic.Create(
+					Diagnostics.CustomServiceLocator,
+					context.Node.GetLocation()
+				)
+			);
+		}
+
+		private static bool IsTestServiceLocatorFactory(
+			INamedTypeSymbol actualType,
+			INamedTypeSymbol disallowedType
+		) {
+			if( actualType == null ) {
+				return false;
+			}
+
+			bool isLocatorFactory = actualType.Equals( disallowedType );
+			return isLocatorFactory;
+		}
+
+		private static bool IsCreateMethod(
+			IMethodSymbol method
+		) {
+			return method.Name.Equals( "Create" )
+				&& method.Parameters.Length > 0;
+		}
+
+		private bool IsClassWhitelisted(
+			ImmutableHashSet<string> whitelistedClasses,
+			ISymbol classSymbol
+		) {
+			bool isWhiteListed = _excludeKnownProblems
+				&& whitelistedClasses.Contains( classSymbol.ToString() );
+
+			return isWhiteListed;
+		}
+
+		private ImmutableHashSet<string> GetWhitelist(
+			ImmutableArray<AdditionalText> additionalFiles
+		) {
+			ImmutableHashSet<string>.Builder whitelistedClasses = ImmutableHashSet.CreateBuilder<string>();
+
+			AdditionalText whitelistFile = additionalFiles.FirstOrDefault(
+				file => Path.GetFileName( file.Path ) == WhitelistFileName
+			);
+
+			if( whitelistFile == null ) {
+				return whitelistedClasses.ToImmutableHashSet();
+			}
+
+			SourceText whitelistText = whitelistFile.GetText();
+
+			foreach( TextLine line in whitelistText.Lines ) {
+				whitelistedClasses.Add( line.ToString().Trim() );

```

Comment:
I figure the `Trim()` here can't hurt. Feedback on adding bad whitespace in here would be almost instantaneous anyways, but it can still be a pretty irritating problem to track down if you're not looking for it.